gax-grpc/src/main/java/com/google/api/gax/grpc/GrpcApiExceptionFactory.java

```diff
@@ -67,8 +84,65 @@ ApiException create(Throwable throwable) {
     boolean canRetry = retryableCodes.contains(GrpcStatusCode.grpcCodeToStatusCode(statusCode));
-    return ApiExceptionFactory.createException(throwable, GrpcStatusCode.of(statusCode), canRetry);
+    GrpcStatusCode grpcStatusCode = GrpcStatusCode.of(statusCode);
+
+    if (metadata == null) {
+      return ApiExceptionFactory.createException(throwable, grpcStatusCode, canRetry);
+    }
+
+    byte[] bytes = metadata.get(Metadata.Key.of(ERROR_DETAIL_KEY, Metadata.BINARY_BYTE_MARSHALLER));
+    if (bytes == null) {
+      return ApiExceptionFactory.createException(throwable, grpcStatusCode, canRetry);
+    }
+
+    com.google.rpc.Status status;
+    try {
+      status = com.google.rpc.Status.parseFrom(bytes);
+    } catch (InvalidProtocolBufferException e) {
+      return ApiExceptionFactory.createException(throwable, grpcStatusCode, canRetry);
+    }
+
+    return ApiExceptionFactory.createException(
+        throwable, grpcStatusCode, canRetry, buildErrorDetails(status));
+  }
+
+  private ErrorDetails buildErrorDetails(com.google.rpc.Status status) {
+    ErrorDetails.Builder errorDetailsBuilder = ErrorDetails.builder();
+    for (Any detail : status.getDetailsList()) {
+      addErrorDetail(errorDetailsBuilder, detail);
+    }
+    return errorDetailsBuilder.build();
+  }
+
+  private void addErrorDetail(ErrorDetails.Builder errorDetailsBuilder, Any detail) {
+    try {
+      if (detail.is(ErrorInfo.class)) {
+        errorDetailsBuilder.setErrorInfo(detail.unpack(ErrorInfo.class));
+      } else if (detail.is(RetryInfo.class)) {
+        errorDetailsBuilder.setRetryInfo(detail.unpack(RetryInfo.class));
+      } else if (detail.is(DebugInfo.class)) {
+        errorDetailsBuilder.setDebugInfo(detail.unpack(DebugInfo.class));
+      } else if (detail.is(QuotaFailure.class)) {
+        errorDetailsBuilder.setQuotaFailure(detail.unpack(QuotaFailure.class));
+      } else if (detail.is(PreconditionFailure.class)) {
+        errorDetailsBuilder.setPreconditionFailure(detail.unpack(PreconditionFailure.class));
+      } else if (detail.is(BadRequest.class)) {
+        errorDetailsBuilder.setBadRequest(detail.unpack(BadRequest.class));
+      } else if (detail.is(RequestInfo.class)) {
+        errorDetailsBuilder.setRequestInfo(detail.unpack(RequestInfo.class));
+      } else if (detail.is(ResourceInfo.class)) {
+        errorDetailsBuilder.setResourceInfo(detail.unpack(ResourceInfo.class));
+      } else if (detail.is(Help.class)) {
+        errorDetailsBuilder.setHelp(detail.unpack(Help.class));
+      } else if (detail.is(LocalizedMessage.class)) {
+        errorDetailsBuilder.setLocalizedMessage(detail.unpack(LocalizedMessage.class));
+      }
+    } catch (InvalidProtocolBufferException e) {
+      // If unpacking one of the error detail fails, it should not block unpacking other error
+      // details so that end users can still get some info back. Hence, we don't need to do
+      // anything.
```

Comment:
If we just swallow this, if details are all corrupt we keep continuing and then create an empty error details and have no insight into what's going on?

Reply:
Yes, in this case, an empty ErrorDetails will be created. I think an empty ErrorDetails is a legitimate case, it could happen either when server doesn't send anything or we are unable to unpack the message server sends.

Reply:
Can we in his case just store the raw Any (not unpacked) `detail` object in `ErrorDetails`? Doing empyt error details will just loose information for users and make the error even more confusing to them.

Reply:
Storing raw Any would not be useful to users in my opinion, they don't know how to unpack it and even if they do they can not unpack it either due to exceptions. It might be more confusing to them because they don't know what it is and don't know what to do with it. 
What I can do though is to add more java docs to this class or the getter to describe what could be null in what scenarios.

Reply:
I would agree they could not unpack it. And just the fact that there is something encoded in any way gives a developer enough clue that there is a speciifc error and they should be able to unpack it sooner or later. It is better than hiding info completely. Another approach may be instead of storing any, convert it to string and store as a string.